File: Static/Passport/Passport/Passport.cpp
#include "Passport.hpp"

#define MAXNUMBER 999999
#define MINNUMBER 100000
#define SERIESLENGTH 2
#define MAXSERIESLETTER 'Z'
#define MINSERIESLETTER 'A'

Passport::Passport(const std::string& firstName, const std::string& secondName, Date* birthDates)
    : m_firstName(firstName), m_secondName(secondName), m_birthDates(birthDates) {

        counter();
        m_series = m_passportSeries;
        m_number = m_passportNumber;
        m_totalPassports += 1;
        m_passports.insert(this);
}

Passport::~Passport() {
    delete m_birthDates;
    m_birthDates = nullptr;

    m_passports.erase(this);
}

const std::string& Passport::getSeries() const {
    return m_series;
}

int Passport::getNumber() const {
    return m_number;
}

const std::string& Passport::getFirstName() const {
    return m_firstName;
}

const std::string& Passport::getSecondName() const {
    return m_secondName;
}

const Date& Passport::getBirthDates() const {
    return *m_birthDates;
}

void Passport::setFirstName(const std::string& newFirstName) {
    m_firstName = newFirstName;
}

void Passport::setSecondName(const std::string& newSecondName) {
    m_secondName = newSecondName;
}

void Passport::setBirthDates(Date* newBirthDates) {
    delete m_birthDates;
    m_birthDates = newBirthDates;
}

std::string Passport::m_passportSeries = SERIESBEGIN;

int Passport::m_passportNumber = STARTNUMBER;

int Passport::m_totalPassports = 0;

std::set<Passport*> Passport::m_passports;

void Passport::counter() {
    if ( m_passportNumber < MAXNUMBER ) {
        m_passportNumber += 1;
    } else {
        m_passportNumber = MINNUMBER;

        if ( m_passportSeries[1] < MAXSERIESLETTER ) {
            m_passportSeries[1] += 1;
        } else {
            m_passportSeries[1] = MINSERIESLETTER;

            if ( m_passportSeries[0] < MAXSERIESLETTER ) {
                m_passportSeries[0] += 1;
            } else {
                throw ReachedLimit();
            }
        }
    }
}

void Passport::verify(const std::string& passportSeries, int passportNumber) {
    if ( passportSeries.length() > SERIESLENGTH ) {
        throw SeriesLengthExceeded();
    }

    if ( !isupper(passportSeries[0]) ) {
        throw InvalidUpperCase();
    } else {
        if ( !isupper(passportSeries[1]) ) {
            throw InvalidUpperCase();
        }
    }

    if ( passportSeries.compare(m_passportSeries) <= 0 ) {
        throw InvalidNewSeries();
    }

    if ( passportNumber > MAXNUMBER || passportNumber < STARTNUMBER) {
        throw InvalidPassportNumber();
    }
}

const std::string& Passport::getPassportSeries() {
    return m_passportSeries;
}

int Passport::getPassportNumber() {
    return m_passportNumber;
}

int Passport::getTotalPassports() {
    return m_totalPassports;
}

void Passport::setNewSeries(const std::string& newPassportSeries, int newPassportNumber) {
    verify(newPassportSeries, newPassportNumber);

    m_passportSeries = newPassportSeries;
    m_passportNumber = newPassportNumber;
}

void Passport::findPassport(Passport* passport) {
    auto it = m_passports.find(passport);

    if ( it == m_passports.end() ) {
        throw PassportNotFound();
    }
    std::cout << *(*it) << std::endl;
}

void Passport::getAllPassports() {
    for ( Passport* passport : m_passports ) {
        std::cout << *passport << std::endl;
    }
}

std::ostream& operator<<(std::ostream& out, const Passport& passport) {
    out << "Passport: " << passport.getSeries() << passport.getNumber() << std::endl
        << "  First Name: " << passport.getFirstName() << ". Second Name: " << passport.getSecondName()<< "." << std::endl
        << "  " << passport.getBirthDates() << std::endl;

    return out;
}

File: PHP/Static/Passport/Passport.php
<?php
    require_once 'CheckFunctions.php';
    
    function exAutoload($className) {
        $path = "Exceptions/$className.php";
        include($path);
    }
    spl_autoload_register("exAutoload");

    define('MAX_NUMBER', 999999);
    define('MIN_NUMBER', 100000);
    define('SERIES_LENGTH', 2);
    define('MAX_SERIES_LETTER', 'Z');
    define('MIN_SERIES_LETTER', 'A');
    
    class Passport {
        private $series, $number;
        private $firstName, $secondName;
        private $birthDate;
        
        private static $passportSeries = 'AA';
        private static $passportNumber = 100000;
        private static $totalPassports = 0;
        private static $passports = array();
        
        private static function counter() {
            $alphabet = array('A','B','C','D','E','F','G','H','I','J','K','L','M',
                'N','O','P','Q','R','S','T','U','V','W','X','Y','Z');
            
            if ( self::$passportNumber < MAX_NUMBER ) {
                self::$passportNumber += 1;
            } else {
                self::$passportNumber = MIN_NUMBER;
                
                if ( self::$passportSeries[1] < MAX_SERIES_LETTER ) {
                    self::$passportSeries[1] = next($alphabet);
                } else {
                    self::$passportSeries[1] = MIN_SERIES_LETTER;
                    
                    if ( self::$passportSeries[0] < MAX_SERIES_LETTER ) {
                        self::$passportSeries[0] = next($alphabet);
                    } else {
                        throw new ReachedLimit("No free variations of the passport series and numbers!");
                    }
                }
            }
        }
        
        public function __construct($firstName, $secondName, $birthDate) {
            $this->firstName = CheckFunctions::isString($firstName);
            $this->secondName = CheckFunctions::isString($secondName);
            $this->birthDate = CheckFunctions::validate($birthDate);
            
            $this->series = self::$passportSeries;
            $this->number = self::$passportNumber;
            
            self::counter();
            self::$totalPassports += 1;
            self::$passports[] = $this;
        }
        
        public function __get($name) {
            if ( property_exists($this, $name) ) {
                return $this->$name;    
            }
            throw new InvalidAttribute("Attribute error: attribute '{$name}' not found.");
        }
        
        public function __set($name, $value) {
            switch ( $name ) {
                case 'firstName':
                    $this->firstName = CheckFunctions::isString($value);    
                    break;
                case 'secondName':
                    $this->secondName = CheckFunctions::isString($value);
                    break;
                case 'birthDate':
                    $this->birthDate = CheckFunctions::validate($value);
                    break;
                default:
                    throw new InvalidAttribute("Attribute error: attribute '{$name}' not found.");
            }
        }
        
        public static function getPassportSeries() {
            return self::$passportSeries;
        }
        
        public static function getPassportNumber() {
            return self::$passportNumber;
        }
        
        public static function getTotalPassports() {
            return self::$totalPassports;    
        }
        
        public static function setNewSeries($newPassportSeries, $newPassportNumber = MIN_NUMBER) {
            CheckFunctions::verify($newPassportSeries, $newPassportNumber);

            self::$passportSeries = strtoupper($newPassportSeries);
            self::$passportNumber = $newPassportNumber;
        }
        
        public static function findPassport(Passport $passport) {
            $key = array_search($passport, self::$passports);

            if ( $key === false ) {
                throw PassportNotFound("Passport was not found.");
            }
            echo self::$passports[$key] . PHP_EOL;
        }
        
        public static function getAllPassports() {
            foreach ( self::$passports as $value ) {
                echo $value . PHP_EOL;
             }
        }
        
        public function __toString() {
            return "Passport: {$this->series}{$this->number}" . PHP_EOL
                . "--{$this->firstName} {$this->secondName}: {$this->birthDate}.";
        }
    }
?>
